### PR TITLE
folder_block_ops: fix SearchForNodes in the case where the node cache can change

### PR DESCRIPTION
**Pull request by strib**
This fixes `TestLongTlfEditHistory`, and also fixes possible issues with conflict resolution.  The general issue is that when `folderBlockOps.SearchForNodes` is passed `folderBranchOps.nodeCache` by some goroutine that doesn't hold any of the `folderBranchOps` locks (like `TlfEditHistory` or `ConflictResolver`), then the pointers in the node cache could be updated by `folderBranchOps` while the search is taking place.

Even holding `folderBlockOps.blockLock` during the search is not enough, because it gets unlocked during network accesses while downloading blocks.

To fix it, this PR does the follow:

* If the cache that's passed into `SearchForNodes` does not contain the expected root node, `SearchForNodes` uses a brand new `NodeCache` instead.  Using a passed-in cache is just an optimization anyway.
  * It also returns the cache it ultimately used, so the caller can use that cache in case it is different than the one it passed in.
* If we finish a search and the root node has a different pointer than we expected, try the entire search again with a fresh cache.
* Add a new `SearchForPaths` method that also gets all the paths while under the lock, so they can't change after `SearchForNodes` finishes.

It also fixes an issue where `folderBranchOps` could end up calling `TlfEditHistory.UpdateHistory` twice for the same MD if it was created locally and then seen again during a race with the background update goroutine.

Issue: KBFS-1365

### CODE REVIEW

**Comment by strib** [comment]
(Note I'm still seeing very rare issues, like 1-in-5000, that I'm still tracking down, so maybe not worth reviewing quite yet.)

**Comment by strib** [comment]
Ready for review!

File: libkbfs/cr_chains.go [hunk 1]
```diff
@@ -864,6 +858,7 @@ func (ccs *crChains) getPaths(ctx context.Context, blocks *folderBlockOps,
 		for _, op := range chain.ops {
 			op.setFinalPath(p)
 		}
+		paths = append(paths, p)
```
Comment (jzila) [inline libkbfs/cr_chains.go]: Just checking: this line presumably moved so that a path is only added if a chain for it exists. Is that correct?

Comment (strib) [inline libkbfs/cr_chains.go]: Oh actually I just ripped apart this function and rewrote it, and an identical line ended up in a different place.  Good catch, I will move it back!

File: libkbfs/folder_block_ops.go [hunk 2]
```diff
@@ -2498,8 +2518,8 @@ func (fbo *folderBlockOps) searchForNodesInDirLocked(ctx context.Context,
 		if _, ok := nodeMap[de.BlockPointer]; ok {
 			childPath := currDir.ChildPath(name, de.BlockPointer)
 			// make a node for every pathnode
-			var n Node
-			for _, pn := range childPath.path {
+			n := rootNode
+			for _, pn := range childPath.path[1:] {
```
Comment (jzila) [inline libkbfs/folder_block_ops.go]: what's going on here? Why are we iterating over `path[1:]` instead of `path`?

Comment (jzila) [inline libkbfs/folder_block_ops.go]: Oh I see, because we already checked the rootNode above.

Comment (strib) [inline libkbfs/folder_block_ops.go]: Yes, exactly.  I figured since we were passing it around now, might as well use it.

File: libkbfs/folder_block_ops.go [hunk 3]
```diff
@@ -2489,6 +2499,16 @@ func (fbo *folderBlockOps) searchForNodesInDirLocked(ctx context.Context,
 		return 0, err
 	}
 
+	// getDirLocked may have unlocked blockLock, which means the cache
+	// could have changed out from under us.  Verify that didn't
+	// happen, so we can avoid messing it up with nodes from an old MD
+	// version.  If it did happen, return a special error that lets
+	// the caller know they should retry with a fresh cache.
```
Comment (jzila) [inline libkbfs/folder_block_ops.go]: Do we need to support multiple simultaneous node caches? If not, we can store it inside folderBlockOps. Additionally, is the problem here that `currDir.path` wouldn't be updated but `cache` would? How can `getDirLocked` update the cache from under us if we didn't pass the `cache` to it?

File: libkbfs/folder_block_ops.go [hunk 4]
```diff
@@ -2551,12 +2566,14 @@ func (fbo *folderBlockOps) SearchForNodes(ctx context.Context,
...
 
-	// Start with the root node
 	rootPtr := md.data.Dir.BlockPointer
 	var node Node
 	if cache == fbo.nodeCache {
-		// Root node should already exist.
+		// Root node should already exist if we have an up-to-date md.
 		node = cache.Get(rootPtr.ref())
+		if node == nil {
+			return nil, searchWithOutOfDateCacheError{}
```
Comment (jzila) [inline libkbfs/folder_block_ops.go]: How can this happen?

Comment (jzila) [inline libkbfs/folder_block_ops.go]: More specifically, how can the cache here become stale?

Comment (jzila) [inline libkbfs/folder_block_ops.go]: Nvmind I see now

File: libkbfs/folder_block_ops.go (hunk 3 above)
Comment (jzila) [inline libkbfs/folder_block_ops.go]: Nvmind, I see now that we have `fbo.nodeCache`, and we pass in a cache to allow attempting the search with a clean cache.

File: libkbfs/folder_block_ops.go [hunk 5]
```diff
@@ -2551,12 +2566,14 @@ func (fbo *folderBlockOps) SearchForNodes(ctx context.Context,
 		return nodeMap, nil
 	}
 
-	// Start with the root node
 	rootPtr := md.data.Dir.BlockPointer
 	var node Node
 	if cache == fbo.nodeCache {
```
Comment (jzila) [inline libkbfs/folder_block_ops.go]: I don't love having these specific corner cases that effectively test for cache invalidation. It seems like it could be pretty brittle. Can we instead actively invalidate `fbo.nodeCache`, so we don't have to deal with these specific checks in cache accesses?

File: libkbfs/folder_block_ops.go [hunk 6]
```diff
@@ -2587,16 +2604,97 @@ func (fbo *folderBlockOps) SearchForNodes(ctx context.Context,
...
+	nodeMap, err := fbo.trySearchWithCacheLocked(ctx, lState, cache, ptrs,
+		newPtrs, md)
+	if _, ok := err.(searchWithOutOfDateCacheError); ok {
+		// The md is out-of-date, so use a throwaway cache so we
+		// don't pollute the real node cache with stale nodes.
+		fbo.log.CDebugf(ctx, "Root node %v doesn't exist in the node "+
+			"cache; using a throwaway node cache instead",
+			md.data.Dir.BlockPointer)
+		cache = newNodeCacheStandard(fbo.folderBranch)
+		nodeMap, err = fbo.trySearchWithCacheLocked(ctx, lState, cache, ptrs,
```
Comment (jzila) [inline libkbfs/folder_block_ops.go]: Because of the recursive calls to `searchForNodesInDirLocked`, there are multiple potential unlocks as part of the path tree traversal. I don't have a solid mental model of how this works yet, but I suspect that even a clean cache might get invalidated by writes while we're traversing for search. WDYT?

**Comment by jzila** [comment]
First pass done. Some questions outstanding.

File: libkbfs/folder_block_ops.go (hunk 5 above)
Comment (strib) [inline libkbfs/folder_block_ops.go]: So the node cache used by the main part of KBFS is `fbo.nodeCache`.  This basically maps from `BlockPointer`s to `Node`s.  `Node`s are used by the callers of the library, but internally we need to know the series of `BlockPointer`s and file/dir names that make up the `path` of the corresponding file/dir.  `fbo.nodeCache` is long-lived and never invalidated.

As `folderBranchOps` gets informed of new local or remote MD updates, which change the `BlockPointer`s of some subset of the nodes in this TLF, it calls `nodeCache.UpdatePointer` for each change.  Then, when a caller passes some old `Node` they have lying around into an FBO call, we can translate it to its current `path` using `fbo.nodeCache`.  Note that on every TLF modification, we are guaranteed that the `BlockPointer` of the root directory will change (because of the merkle-ish tree of content hashes we use to assign `BlockPointer`s).

That's a long way of saying we never "invalidate" `fbo.nodeCache`.  It needs to maintain the absolute latest mappings for the TLF, or else FBO calls won't see up-to-date data.  The tension in `SearchForNodes` comes from the fact that we are trying to discover the `BlockPointer`s of certain files _at a specific point in the MD history_, which is not necessarily the same as the most-recently-seen MD update.  Specifically, for the `TlfEditHistory` we are processing a specific range of MDs, but `folderBranchOps` may have heard about a newer one before, or during, `TlfEditHistory` started processing.  That means `fbo.nodeCache` may have been updated to reflect the newest `BlockPointer`s, and is no longer correct as a cache for our search for the data at the old point in time.

So, we basically have two options:
1. Figure out if `fbo.nodeCache` still represents the state of the system at the point of time we care about.  That's what we're doing here by checking that the root pointer for the MD update we care about is still in the cache; if there had been an update in the meantime, that root pointer would have changed.
2. Never search using `fbo.nodeCache`, and always pass in a clean cache.  This would simplify the code, but having a populated node cache allows us to search a bit faster; plus the nodes we create may be useful to other parts of the system later on.

So, that's the background here, and why this code special cases `fbo.nodeCache`.  I'm not sure if that addresses your concern, because I don't know if I fully get what you mean when you say "invalidate `fbo.nodeCache`".

File: libkbfs/folder_block_ops.go (hunk 6 above)
Comment (strib) [inline libkbfs/folder_block_ops.go]: We are guaranteed that the clean cache will not have its root pointer changed, because there are no pointers to it outside of this flow of execution.  We create it here, and it never gets passed to anything that modifies it.  (It's not passed to `getDirLocked`, and it's not saved in any structs, etc.

Does that help?  Is there something I can do to make it more clear?

File: libkbfs/folder_block_ops.go (hunk 5 above)
Comment (jzila) [inline libkbfs/folder_block_ops.go]: Yeah I wrote this comment before I realized why the caches were being passed in. Your explanation here, however, is fantastic, and it might be a good idea to store it somewhere for posterity. The key thing to notice is that `fbo.nodeCache` is the _latest_ view, but the search needs a view fixed in time.

File: libkbfs/folder_block_ops.go (hunk 6 above)
Comment (jzila) [inline libkbfs/folder_block_ops.go]: Your explanation above was key, I understand now. Thanks Jeremy!

**Comment by jzila** [comment]
LGTM!

**Comment by akalin-keybase** [comment]
lgtm!

File: libkbfs/folder_block_ops.go (hunk 5 above)
Comment (strib) [inline libkbfs/folder_block_ops.go]: Good idea, I've pasted most of the above comment above the search retry logic.  Not sure if that's the best place, but put it there for now so we don't lose track of it.
